File: Practice/Walmart/karat/Domain Analysis.py
"""
The "Domain Analysis" problem from Karat assesses your ability to understand and analyze a given problem domain, focusing on identifying key components, relationships, and requirements. This skill is crucial for designing effective software solutions.

Problem Statement:

You are tasked with designing a system for a library management application. The system should handle the following functionalities:

Book Management:
Add new books to the catalog.
Update book information (e.g., title, author, genre).
Remove books from the catalog.
User Management:
Register new users.
Update user information (e.g., contact details).
Deactivate user accounts.
Loan Management:
Allow users to borrow books.
Track due dates and overdue books.
Handle book returns and renewals.
Search Functionality:
Enable users to search for books by title, author, or genre.
Provide recommendations based on user preferences and borrowing history.
Your Task:

Conduct a domain analysis to identify the key entities, their attributes, and the relationships between them. Based on your analysis, design a class diagram that represents the structure of the system. Additionally, outline the main use cases and describe the interactions between users and the system for each functionality.

Considerations:

Think about the data types and constraints for each attribute.
Identify any potential exceptions or error conditions.
Consider scalability and performance requirements.
Ensure that the design supports future extensions, such as adding new functionalities or integrating with other systems.

"""


from typing import List
from datetime import datetime

class Book:
    def __init__(self, id: int, title: str, author: str, genre: str):
        self.id = id
        self.title = title
        self.author = author
        self.genre = genre
        self.status = "available"
        self.due_date = None

    def update_details(self, title: str, author: str, genre: str):
        self.title = title
        self.author = author
        self.genre = genre

    def borrow(self):
        self.status = "borrowed"

    def return_book(self):
        self.status = "available"
        self.due_date = None

    def renew(self):
        # Extend due date for another week, for example
        self.due_date = datetime.now() + timedelta(days=7)

class User:
    def __init__(self, id: int, name: str, email: str):
        self.id = id
        self.name = name
        self.email = email
        self.status = "active"

    def register(self):
        # Register a new user
        pass

    def update_info(self, name: str, email: str):
        self.name = name
        self.email = email

    def deactivate_account(self):
        self.status = "inactive"

    def borrow_book(self, book: Book):
        book.borrow()

    def return_book(self, book: Book):
        book.return_book()

    def renew_book(self, book: Book):
        book.renew()

class Loan:
    def __init__(self, user_id: int, book_id: int, loan_date: datetime, due_date: datetime):
        self.user_id = user_id
        self.book_id = book_id
        self.loan_date = loan_date
        self.due_date = due_date

    def extend_due_date(self):
        self.due_date += timedelta(days=7)

class Library:
    def __init__(self):
        self.books = []
        self.users = []
        self.loans = []

    def add_book(self, book: Book):
        self.books.append(book)

    def remove_book(self, book: Book):
        self.books.remove(book)

    def search_books(self, query: str):
        return [book for book in self.books if query.lower() in book.title.lower() or query.lower() in book.author.lower()]

    def recommend_books(self, user: User):
        # Implement a simple recommendation logic based on past loans
        return self.books  # A placeholder for actual recommendation logic

File: Practice/Walmart/karat/Domain analysis.java
import java.util.*;

class Book {
    private int id;
    private String title;
    private String author;
    private String genre;
    private String status;  // "available", "borrowed"
    private Date dueDate;

    public Book(int id, String title, String author, String genre) {
        this.id = id;
        this.title = title;
        this.author = author;
        this.genre = genre;
        this.status = "available";
        this.dueDate = null;
    }

    // Getter and Setter methods
    public int getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public String getAuthor() {
        return author;
    }

    public String getGenre() {
        return genre;
    }

    public String getStatus() {
        return status;
    }

    public Date getDueDate() {
        return dueDate;
    }

    // Methods to update book details and manage its state
    public void updateDetails(String title, String author, String genre) {
        this.title = title;
        this.author = author;
        this.genre = genre;
    }

    public void borrow() {
        this.status = "borrowed";
    }

    public void returnBook() {
        this.status = "available";
        this.dueDate = null;
    }

    public void renew() {
        // Renew the book (extend due date)
        this.dueDate = new Date(System.currentTimeMillis() + 7 * 24 * 60 * 60 * 1000L);  // Add 7 days to the current time
    }
}

class User {
    private int id;
    private String name;
    private String email;
    private String status;  // "active", "inactive"

    public User(int id, String name, String email) {
        this.id = id;
        this.name = name;
        this.email = email;
        this.status = "active";
    }

    // Getter and Setter methods
    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }

    public String getEmail() {
        return email;
    }

    public String getStatus() {
        return status;
    }

    // Methods to manage user actions
    public void register() {
        // Code to register user
    }

    public void updateInfo(String name, String email) {
        this.name = name;
        this.email = email;
    }

    public void deactivateAccount() {
        this.status = "inactive";
    }

    public void borrowBook(Book book) {
        book.borrow();
    }

    public void returnBook(Book book) {
        book.returnBook();
    }

    public void renewBook(Book book) {
        book.renew();
    }
}

class Loan {
    private int userId;
    private int bookId;
    private Date loanDate;
    private Date dueDate;

    public Loan(int userId, int bookId, Date loanDate, Date dueDate) {
        this.userId = userId;
        this.bookId = bookId;
        this.loanDate = loanDate;
        this.dueDate = dueDate;
    }

    // Getter and Setter methods
    public int getUserId() {
        return userId;
    }

    public int getBookId() {
        return bookId;
    }

    public Date getLoanDate() {
        return loanDate;
    }

    public Date getDueDate() {
        return dueDate;
    }

    // Extend the due date for the loan
    public void extendDueDate() {
        this.dueDate = new Date(this.dueDate.getTime() + 7 * 24 * 60 * 60 * 1000L);  // Extend 7 more days
    }
}

class Library {
    private List<Book> books;
    private List<User> users;
    private List<Loan> loans;

    public Library() {
        this.books = new ArrayList<>();
        this.users = new ArrayList<>();
        this.loans = new ArrayList<>();
    }

    // Methods to manage books
    public void addBook(Book book) {
        books.add(book);
    }

    public void removeBook(Book book) {
        books.remove(book);
    }

    public List<Book> searchBooks(String query) {
        List<Book> result = new ArrayList<>();
        for (Book book : books) {
            if (book.getTitle().toLowerCase().contains(query.toLowerCase()) || book.getAuthor().toLowerCase().contains(query.toLowerCase())) {
                result.add(book);
            }
        }
        return result;
    }

    // Recommend books to the user (this can be improved with more sophisticated logic)
    public List<Book> recommendBooks(User user) {
        // For simplicity, we recommend all available books
        List<Book> recommendedBooks = new ArrayList<>();
        for (Book book : books) {
            if (book.getStatus().equals("available")) {
                recommendedBooks.add(book);
            }
        }
        return recommendedBooks;
    }
}
